File: rclpy/rclpy/action_client.py
# Copyright 2019 Open Source Robotics Foundation, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# import threading
import time
import uuid

from action_msgs.msg import GoalStatus
from action_msgs.srv import CancelGoal

from rclpy.impl.implementation_singleton import rclpy_action_implementation as _rclpy_action
# TODO(jacobperron): Move check_for_type_support to its own module (e.g. type_support)
#                    Do after Crystal patch release since this breaks API
from rclpy.node import check_for_type_support
from rclpy.qos import qos_profile_action_status_default
from rclpy.qos import qos_profile_default, qos_profile_services_default
from rclpy.task import Future
from rclpy.waitable import Waitable

from unique_identifier_msgs.msg import UUID


class ClientGoalHandle():
    """Goal handle for working with Action Clients."""

    def __init__(self, goal_id, goal_response):
        if not isinstance(goal_id, UUID):
            raise TypeError('Expected UUID, but given {}'.format(type(goal_id)))

        self._goal_id = goal_id
        self._goal_response = goal_response
        self._status = GoalStatus.STATUS_UNKNOWN

    def __eq__(self, other):
        return self._goal_id == other.goal_id

    def __repr__(self):
        return 'ClientGoalHandle <{}>'.format(self.goal_id.uuid)

    @property
    def goal_id(self):
        return self._goal_id

    @property
    def goal_response(self):
        return self._goal_response

    @property
    def accepted(self):
        return self._goal_response.accepted

    @property
    def status(self):
        return self._status


class ActionClient(Waitable):
    """ROS Action client."""

    def __init__(
        self,
        node,
        action_type,
        action_name,
        *,
        callback_group=None,
        goal_service_qos_profile=qos_profile_services_default,
        result_service_qos_profile=qos_profile_services_default,
        cancel_service_qos_profile=qos_profile_services_default,
        feedback_sub_qos_profile=qos_profile_default,
        status_sub_qos_profile=qos_profile_action_status_default
    ):
        """
        Constructor.

        :param node: The ROS node to add the action client to.
        :param action_type: Type of the action.
        :param action_name: Name of the action.
            Used as part of the underlying topic and service names.
        :param callback_group: Callback group to add the action client to.
            If None, then the node's default callback group is used.
        :param goal_service_qos_profile: QoS profile for the goal service.
        :param result_service_qos_profile: QoS profile for the result service.
        :param cancel_service_qos_profile: QoS profile for the cancel service.
        :param feedback_sub_qos_profile: QoS profile for the feedback subscriber.
        :param status_sub_qos_profile: QoS profile for the status subscriber.
        """
        if callback_group is None:
            callback_group = node.default_callback_group

        super().__init__(callback_group)

        # Import the typesupport for the action module if not already done
        check_for_type_support(action_type)
        self.node = node
        self.action_type = action_type
        self.client_handle = _rclpy_action.rclpy_action_create_client(
            node.handle,
            action_type,
            action_name,
            goal_service_qos_profile,
            result_service_qos_profile,
            cancel_service_qos_profile,
            feedback_sub_qos_profile,
            status_sub_qos_profile
        )

        self._is_ready = False
        self._goal_handles = {}
        self._pending_goal_requests = {}
        self._sequence_number_to_goal_id = {}  # goal request sequence number
        self._pending_cancel_requests = {}
        self._pending_result_requests = {}
        self._feedback_callbacks = {}

        callback_group.add_entity(self)
        self.node.add_waitable(self)

    def _generate_random_uuid(self):
        return UUID(uuid=list(uuid.uuid4().bytes))

    def _remove_pending_request(self, future, pending_requests):
        """
        Remove a future from the list of pending requests.

        This prevents a future from receiving a request and executing its done callbacks.
        :param future: a future returned from one of :meth:`send_goal_async`,
            :meth:`cancel_goal_async`, or :meth:`get_result_async`.
        :type future: rclpy.task.Future
        :param pending_requests: The list of pending requests.
        :type pending_requests: dict
        :return: The sequence number associated with the removed future, or
            None if the future was not found in the list.
        """
        for seq, req_future in pending_requests.items():
            if future == req_future:
                try:
                    del pending_requests[seq]
                except KeyError:
                    pass
                return seq
        return None

    def _remove_pending_goal_request(self, future):
        seq = self._remove_pending_request(future, self._pending_goal_requests)
        if seq in self._sequence_number_to_goal_id:
            del self._sequence_number_to_goal_id[seq]

    def _remove_pending_cancel_request(self, future):
        self._remove_pending_request(future, self._pending_cancel_requests)

    def _remove_pending_result_request(self, future):
        self._remove_pending_request(future, self._pending_result_requests)

    # Start Waitable API
    def is_ready(self, wait_set):
        """Return True if one or more entities are ready in the wait set."""
        ready_entities = _rclpy_action.rclpy_action_wait_set_is_ready(self.client_handle, wait_set)
        self._is_feedback_ready = ready_entities[0]
        self._is_status_ready = ready_entities[1]
        self._is_goal_response_ready = ready_entities[2]
        self._is_cancel_response_ready = ready_entities[3]
        self._is_result_response_ready = ready_entities[4]
        return any(ready_entities)

    def take_data(self):
        """Take stuff from lower level so the wait set doesn't immediately wake again."""
        data = {}
        if self._is_feedback_ready:
            feedback_msg = _rclpy_action.rclpy_action_take_feedback(
                self.client_handle, self.action_type.Feedback)
            data['feedback'] = feedback_msg

        if self._is_status_ready:
            status_msg = _rclpy_action.rclpy_action_take_status(
                self.client_handle, self.action_type.GoalStatusMessage)
            data['status'] = status_msg

        if self._is_goal_response_ready:
            sequence_number, goal_response = _rclpy_action.rclpy_action_take_goal_response(
                self.client_handle, self.action_type.GoalRequestService.Response)
            data['goal'] = (sequence_number, goal_response)

        if self._is_cancel_response_ready:
            sequence_number, cancel_response = _rclpy_action.rclpy_action_take_cancel_response(
                self.client_handle, self.action_type.CancelGoalService.Response)
            data['cancel'] = (sequence_number, cancel_response)

        if self._is_result_response_ready:
            sequence_number, result_response = _rclpy_action.rclpy_action_take_result_response(
                self.client_handle, self.action_type.GoalResultService.Response)
            data['result'] = (sequence_number, result_response)

        if not any(data):
            return None
        return data

    async def execute(self, taken_data):
        """
        Execute work after data has been taken from a ready wait set.

        This will set results for Future objects for any received service responses and
        call any user-defined callbacks (e.g. feedback).
        """
        if 'feedback' in taken_data:
            feedback_msg = taken_data['feedback']
            goal_uuid = uuid.UUID(bytes=bytes(feedback_msg.action_goal_id.uuid))
            # Call a registered callback if there is one
            if goal_uuid in self._feedback_callbacks:
                self._feedback_callbacks[goal_uuid](feedback_msg)

        if 'status' in taken_data:
            # Update the status of all goal handles maintained by this Action Client
            for status_msg in taken_data['status'].status_list:
                goal_uuid = bytes(status_msg.goal_info.goal_id.uuid)
                status = status_msg.status
                if goal_uuid in self._goal_handles:
                    self._goal_handles[goal_uuid]._status = status
                    # Remove "done" goals from the list
                    if (GoalStatus.STATUS_SUCCEEDED == status or
                            GoalStatus.STATUS_CANCELED == status or
                            GoalStatus.STATUS_ABORTED == status):
                        del self._goal_handles[goal_uuid]

        if 'goal' in taken_data:
            sequence_number, goal_response = taken_data['goal']
            goal_handle = ClientGoalHandle(
                self._sequence_number_to_goal_id[sequence_number],
                goal_response)

            if goal_handle.accepted:
                goal_uuid = bytes(goal_handle.goal_id.uuid)
                if goal_uuid in self._goal_handles:
                    raise RuntimeError(
                        'Two goals were accepted with the same ID ({})'.format(goal_handle))
                self._goal_handles[goal_uuid] = goal_handle

            self._pending_goal_requests[sequence_number].set_result(goal_handle)

        if 'cancel' in taken_data:
            sequence_number, cancel_response = taken_data['cancel']
            self._pending_cancel_requests[sequence_number].set_result(cancel_response)

        if 'result' in taken_data:
            sequence_number, result_response = taken_data['result']
            self._pending_result_requests[sequence_number].set_result(result_response)

    def get_num_entities(self):
        """Return number of each type of entity used in the wait set."""
        return _rclpy_action.rclpy_action_wait_set_get_num_entities(self.client_handle)

    def add_to_wait_set(self, wait_set):
        """Add entities to wait set."""
        _rclpy_action.rclpy_action_wait_set_add(self.client_handle, wait_set)
    # End Waitable API

    def send_goal(self, goal, **kwargs):
        """
        Send a goal and wait for the result.

        Do not call this method in a callback or a deadlock may occur.

        See :meth:`send_goal_async` for more info about keyword arguments.

        :param goal: The goal request.
        :type goal: action_type.Goal
        :return: The result response.
        :rtype: action_type.Result
        """
        future = self.send_goal_async(goal, kwargs)
        while self.node.context.ok() and not future.done():
            time.sleep(0.1)
        return future.result()

    def send_goal_async(self, goal, feedback_callback=None, goal_uuid=None):
        """
        Send a goal and asynchronously get the result.

        The result of the returned Future is set to a ClientGoalHandle when receipt of the goal
        is acknowledged by an action server.

        :param goal: The goal request.
        :type goal: action_type.Goal
        :param feedback_callback: Callback function for feedback associated with the goal.
        :type feedback_callback: function
        :param goal_uuid: Universally unique identifier for the goal.
            If None, then a random UUID is generated.
        :type: unique_identifier_msgs.UUID
        :return: a Future instance to a goal handle that completes when the goal request
            has been accepted or rejected.
        :rtype: :class:`rclpy.task.Future` instance
        """
        goal.action_goal_id = self._generate_random_uuid() if goal_uuid is None else goal_uuid
        sequence_number = _rclpy_action.rclpy_action_send_goal_request(self.client_handle, goal)
        if sequence_number in self._pending_goal_requests:
            raise RuntimeError(
                'Sequence ({}) conflicts with pending goal request'.format(sequence_number))

        if feedback_callback is not None:
            # TODO(jacobperron): Move conversion function to a general-use package
            goal_uuid = uuid.UUID(bytes=bytes(goal.action_goal_id.uuid))
            self._feedback_callbacks[goal_uuid] = feedback_callback

        future = Future()
        self._pending_goal_requests[sequence_number] = future
        self._sequence_number_to_goal_id[sequence_number] = goal.action_goal_id
        future.add_done_callback(self._remove_pending_goal_request)

        return future

    def cancel_goal(self, goal_handle):
        """
        Send a cancel request for an active goal and wait for the response.

        Do not call this method in a callback or a deadlock may occur.

        :param goal_handle: Handle to the goal to cancel.
        :type goal_handle: :class:`ClientGoalHandle`
        :return: The cancel response.
        """
        pass

    def cancel_goal_async(self, goal_handle):
        """
        Send a cancel request for an active goal and asynchronously get the result.

        :param goal_handle: Handle to the goal to cancel.
        :type goal_handle: :class:`ClientGoalHandle`
        :return: a Future instance that completes when the cancel request has been processed.
        :rtype: :class:`rclpy.task.Future` instance
        """
        if not isinstance(goal_handle, ClientGoalHandle):
            raise TypeError(
                'Expected type ClientGoalHandle but received {}'.format(type(goal_handle)))

        cancel_request = CancelGoal.Request()
        cancel_request.goal_info.goal_id = goal_handle.goal_id
        sequence_number = _rclpy_action.rclpy_action_send_cancel_request(
            self.client_handle,
            cancel_request)
        if sequence_number in self._pending_cancel_requests:
            raise RuntimeError(
                'Sequence ({}) conflicts with pending cancel request'.format(sequence_number))

        future = Future()
        self._pending_cancel_requests[sequence_number] = future
        future.add_done_callback(self._remove_pending_cancel_request)

        return future

    def get_result(self, goal_handle):
        """
        Request the result for an active goal and wait for the response.

        Do not call this method in a callback or a deadlock may occur.

        :param goal_handle: Handle to the goal to get the result for.
        :type goal_handle: :class:`ClientGoalHandle`
        :return: The result response.
        """
        pass

    def get_result_async(self, goal_handle):
        """
        Request the result for an active goal asynchronously.

        :param goal_handle: Handle to the goal to cancel.
        :type goal_handle: :class:`ClientGoalHandle`
        :return: a Future instance that completes when the get result request has been processed.
        :rtype: :class:`rclpy.task.Future` instance
        """
        if not isinstance(goal_handle, ClientGoalHandle):
            raise TypeError(
                'Expected type ClientGoalHandle but received {}'.format(type(goal_handle)))

        result_request = self.action_type.GoalResultService.Request()
        result_request.action_goal_id = goal_handle.goal_id
        sequence_number = _rclpy_action.rclpy_action_send_result_request(
            self.client_handle,
            result_request)
        if sequence_number in self._pending_result_requests:
            raise RuntimeError(
                'Sequence ({}) conflicts with pending result request'.format(sequence_number))

        future = Future()
        self._pending_result_requests[sequence_number] = future
        future.add_done_callback(self._remove_pending_result_request)

        return future

    def server_is_ready(self):
        """
        Check if there is an action server ready to process requests from this client.

        :return: True if an action server is ready, False otherwise.
        """
        return _rclpy_action.rclpy_action_server_is_available(self.node.handle, self.client_handle)

    def wait_for_server(self, timeout_sec=None):
        """
        Wait for an action sever to be ready.

        Returns as soon as an action server is ready for this client.
        :param timeout_sec: Number of seconds to wait until an action server is available.
            If None, then wait indefinitely.
        :return: True if an action server is available, False if the timeout is exceeded.
        """
        # TODO(jacobperron): Remove arbitrary sleep time and return as soon as server is ready
        #                    See https://github.com/ros2/rclpy/issues/58
        sleep_time = 0.25
        if timeout_sec is None:
            timeout_sec = float('inf')
        while self.node.context.ok() and not self.server_is_ready() and timeout_sec > 0.0:
            time.sleep(sleep_time)
            timeout_sec -= sleep_time

        return self.server_is_ready()

    def destroy(self):
        """Destroy the underlying action client handle."""
        if self.client_handle is None:
            return
        _rclpy_action.rclpy_action_destroy_entity(self.client_handle, self.node.handle)
        self.node.remove_waitable(self)
        self.client_handle = None

    def __del__(self):
        """Destroy the underlying action client handle."""
        self.destroy()

File: rclpy/src/rclpy/action_client.cpp
// Copyright 2021 Open Source Robotics Foundation, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <pybind11/pybind11.h>

#include <rcl/error_handling.h>
#include <rcl/types.h>
#include <rcl_action/action_client.h>
#include <rcl_action/wait.h>
#include <rosidl_runtime_c/action_type_support_struct.h>
#include <rmw/types.h>

#include <memory>
#include <string>

#include "action_client.hpp"
#include "exceptions.hpp"
#include "node.hpp"
#include "utils.hpp"

namespace rclpy
{

void
ActionClient::destroy()
{
  rcl_action_client_.reset();
  node_.destroy();
}

ActionClient::ActionClient(
  Node & node,
  py::object pyaction_type,
  const char * action_name,
  const rmw_qos_profile_t & goal_service_qos,
  const rmw_qos_profile_t & result_service_qos,
  const rmw_qos_profile_t & cancel_service_qos,
  const rmw_qos_profile_t & feedback_topic_qos,
  const rmw_qos_profile_t & status_topic_qos)
: node_(node)
{
  rosidl_action_type_support_t * ts =
    static_cast<rosidl_action_type_support_t *>(common_get_type_support(pyaction_type));
  if (!ts) {
    throw py::error_already_set();
  }

  rcl_action_client_options_t action_client_ops = rcl_action_client_get_default_options();

  action_client_ops.goal_service_qos = goal_service_qos;
  action_client_ops.result_service_qos = result_service_qos;
  action_client_ops.cancel_service_qos = cancel_service_qos;
  action_client_ops.feedback_topic_qos = feedback_topic_qos;
  action_client_ops.status_topic_qos = status_topic_qos;

  rcl_action_client_ = std::shared_ptr<rcl_action_client_t>(
    new rcl_action_client_t,
    [node](rcl_action_client_t * action_client)
    {
      // Intentionally capture node by value so shared_ptr can be transferred to copies
      rcl_ret_t ret = rcl_action_client_fini(action_client, node.rcl_ptr());
      if (RCL_RET_OK != ret) {
        // Warning should use line number of the current stack frame
        int stack_level = 1;
        PyErr_WarnFormat(
          PyExc_RuntimeWarning, stack_level, "Failed to fini publisher: %s",
          rcl_get_error_string().str);
        rcl_reset_error();
      }
      delete action_client;
    });

  *rcl_action_client_ = rcl_action_get_zero_initialized_client();

  rcl_ret_t ret = rcl_action_client_init(
    rcl_action_client_.get(),
    node_.rcl_ptr(),
    ts,
    action_name,
    &action_client_ops);
  if (RCL_RET_ACTION_NAME_INVALID == ret) {
    std::string error_text{"Failed to create action client due to invalid topic name '"};
    error_text += action_name;
    error_text += "' : ";
    error_text += rcl_get_error_string().str;
    rcl_reset_error();
    throw py::value_error(error_text);
  }
  if (RCL_RET_OK != ret) {
    std::string error_text{"Failed to create action client: "};
    error_text += rcl_get_error_string().str;
    rcl_reset_error();
    throw py::value_error(error_text);
  }
}

#define TAKE_SERVICE_RESPONSE(Type) \
  /* taken_msg is always destroyed in this function */ \
  auto taken_msg = create_from_py(pymsg_type); \
  rmw_request_id_t header; \
  rcl_ret_t ret = rcl_action_take_ ## Type ## _response( \
    rcl_action_client_.get(), &header, taken_msg.get()); \
  int64_t sequence = header.sequence_number; \
  /* Create the tuple to return */ \
  if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret || RCL_RET_ACTION_SERVER_TAKE_FAILED == ret) { \
    return py::make_tuple(py::none(), py::none()); \
  } else if (RCL_RET_OK != ret) { \
    throw rclpy::RCLError("Failed to take " #Type); \
  } \
  return py::make_tuple(sequence, convert_to_py(taken_msg.get(), pymsg_type)); \

py::tuple
ActionClient::take_goal_response(py::object pymsg_type)
{
  TAKE_SERVICE_RESPONSE(goal)
}

#define SEND_SERVICE_REQUEST(Type) \
  auto ros_request = convert_from_py(pyrequest); \
  int64_t sequence_number; \
  rcl_ret_t ret = rcl_action_send_ ## Type ## _request( \
    rcl_action_client_.get(), ros_request.get(), &sequence_number); \
  if (RCL_RET_OK != ret) { \
    throw rclpy::RCLError("Failed to send " #Type " request"); \
  } \
  return sequence_number;

int64_t
ActionClient::send_result_request(py::object pyrequest)
{
  SEND_SERVICE_REQUEST(result);
}

py::tuple
ActionClient::take_cancel_response(py::object pymsg_type)
{
  TAKE_SERVICE_RESPONSE(cancel)
}

#define TAKE_MESSAGE(Type) \
  auto taken_msg = create_from_py(pymsg_type); \
  rcl_ret_t ret = rcl_action_take_ ## Type(rcl_action_client_.get(), taken_msg.get()); \
  if (RCL_RET_OK != ret) { \
    if (RCL_RET_ACTION_CLIENT_TAKE_FAILED == ret) { \
      /* if take failed, just do nothing */ \
      return py::none(); \
    } \
    throw rclpy::RCLError("Failed to take " #Type " with an action client"); \
  } \
  return convert_to_py(taken_msg.get(), pymsg_type);

py::object
ActionClient::take_feedback(py::object pymsg_type)
{
  TAKE_MESSAGE(feedback)
}

py::object
ActionClient::take_status(py::object pymsg_type)
{
  TAKE_MESSAGE(status)
}

int64_t
ActionClient::send_cancel_request(py::object pyrequest)
{
  SEND_SERVICE_REQUEST(cancel)
}

int64_t
ActionClient::send_goal_request(py::object pyrequest)
{
  SEND_SERVICE_REQUEST(goal)
}

py::tuple
ActionClient::take_result_response(py::object pymsg_type)
{
  TAKE_SERVICE_RESPONSE(result);
}

py::tuple
ActionClient::get_num_entities()
{
  size_t num_subscriptions = 0u;
  size_t num_guard_conditions = 0u;
  size_t num_timers = 0u;
  size_t num_clients = 0u;
  size_t num_services = 0u;

  rcl_ret_t ret;
  ret = rcl_action_client_wait_set_get_num_entities(
    rcl_action_client_.get(),
    &num_subscriptions,
    &num_guard_conditions,
    &num_timers,
    &num_clients,
    &num_services);
  if (RCL_RET_OK != ret) {
    std::string error_text{"Failed to get number of entities for 'rcl_action_client_t'"};
    throw rclpy::RCLError(error_text);
  }

  return py::make_tuple(
    num_subscriptions, num_guard_conditions, num_timers,
    num_clients, num_services);
}

bool
ActionClient::is_action_server_available()
{
  bool is_available = false;
  rcl_ret_t ret = rcl_action_server_is_available(
    node_.rcl_ptr(), rcl_action_client_.get(), &is_available);
  if (RCL_RET_OK != ret) {
    throw rclpy::RCLError("Failed to check if action server is available");
  }
  return is_available;
}

void
ActionClient::add_to_waitset(WaitSet & wait_set)
{
  rcl_ret_t ret = rcl_action_wait_set_add_action_client(
    wait_set.rcl_ptr(), rcl_action_client_.get(), NULL, NULL);
  if (RCL_RET_OK != ret) {
    std::string error_text{"Failed to add 'rcl_action_client_t' to wait set"};
    throw rclpy::RCLError(error_text);
  }
}

py::tuple
ActionClient::is_ready(WaitSet & wait_set)
{
  bool is_feedback_ready = false;
  bool is_status_ready = false;
  bool is_goal_response_ready = false;
  bool is_cancel_response_ready = false;
  bool is_result_response_ready = false;
  rcl_ret_t ret = rcl_action_client_wait_set_get_entities_ready(
    wait_set.rcl_ptr(),
    rcl_action_client_.get(),
    &is_feedback_ready,
    &is_status_ready,
    &is_goal_response_ready,
    &is_cancel_response_ready,
    &is_result_response_ready);
  if (RCL_RET_OK != ret) {
    throw rclpy::RCLError("Failed to get number of ready entities for action client");
  }

  py::tuple result_tuple(5);
  result_tuple[0] = py::bool_(is_feedback_ready);
  result_tuple[1] = py::bool_(is_status_ready);
  result_tuple[2] = py::bool_(is_goal_response_ready);
  result_tuple[3] = py::bool_(is_cancel_response_ready);
  result_tuple[4] = py::bool_(is_result_response_ready);
  return result_tuple;
}

void
define_action_client(py::object module)
{
  py::class_<ActionClient, Destroyable, std::shared_ptr<ActionClient>>(module, "ActionClient")
  .def(
    py::init<Node &, py::object, const char *, const rmw_qos_profile_t &,
    const rmw_qos_profile_t &, const rmw_qos_profile_t &,
    const rmw_qos_profile_t &, const rmw_qos_profile_t &>())
  .def_property_readonly(
    "pointer", [](const ActionClient & action_client) {
      return reinterpret_cast<size_t>(action_client.rcl_ptr());
    },
    "Get the address of the entity as an integer")
  .def(
    "take_goal_response", &ActionClient::take_goal_response,
    "Take an action goal response.")
  .def(
    "send_result_request", &ActionClient::send_result_request,
    "Send an action result request.")
  .def(
    "take_cancel_response", &ActionClient::take_cancel_response,
    "Take an action cancel response.")
  .def(
    "take_feedback", &ActionClient::take_feedback,
    "Take a feedback message from a given action client.")
  .def(
    "send_cancel_request", &ActionClient::send_cancel_request,
    "Send an action cancel request.")
  .def(
    "send_goal_request", &ActionClient::send_goal_request,
    "Send an action goal request.")
  .def(
    "take_result_response", &ActionClient::take_result_response,
    "Take an action result response.")
  .def(
    "get_num_entities", &ActionClient::get_num_entities,
    "Get the number of wait set entities that make up an action entity.")
  .def(
    "is_action_server_available", &ActionClient::is_action_server_available,
    "Check if an action server is available for the given action client.")
  .def(
    "add_to_waitset", &ActionClient::add_to_waitset,
    "Add an action entity to a wait set.")
  .def(
    "is_ready", &ActionClient::is_ready,
    "Check if an action entity has any ready wait set entities.")
  .def(
    "take_status", &ActionClient::take_status,
    "Take an action status response.");
}
}  // namespace rclpy
